lib/rules/base.js

```diff
@@ -48,57 +43,300 @@ module.exports = function(options) {
+        if (elementNode === parentNode) {
+          pluginContext._popConfig(config.value);
+          return false;
+        }
+      });
+    });
+
+    // Handle in-element config statements
+    astHandlers.ElementNode.enter.before.push(function(elementNode) {
+      elementNode.comments.forEach(function(commentNode) {
+        var config = pluginContext._processInstructionNode(commentNode);
+        if (!config) {
+          return;
         }
 
-        commentVisitorEnter.apply(this, arguments);
-      };
-    }
+        pluginContext._pushConfig(config.value);
+
+        if (config.tree) {
+          // Applies to descendants, so pop the config once we exit the
+          // comment's element
+          astHandlers.ElementNode.exit.after.unshift(function(node) {
+            if (node === elementNode) {
+              pluginContext._popConfig(config.value);
+              return false;
+            }
+          });
+        } else {
+          // Applies to only this element, so pop the config when we enter the
+          // element's children, re-push it when we exit the element's
+          // children, and then re-pop it when we exit the element entirely.
+          astHandlers.ElementNode.keys.children.enter.before.unshift(function(node) {
+            if (node === elementNode) {
+              pluginContext._popConfig(config.value);
+              return false;
+            }
+          });
+          astHandlers.ElementNode.keys.children.exit.after.push(function(node) {
+            if (node === elementNode) {
+              pluginContext._pushConfig(config.value);
+              return false;
+            }
+          });
+          astHandlers.ElementNode.exit.after.unshift(function(node) {
+            if (node === elementNode) {
+              pluginContext._popConfig(config.value);
+              return false;
+            }
+          });
+        }
+      });
+    });
 
     return visitors;
   };
 
-  BasePlugin.prototype.processVisitor = function(ruleVisitor) {
-    var pluginContext = this;
-    var visitor = {
-      enter: function(node) {
-        if (!pluginContext.isDisabled()) {
-          invokeVisitorEnter(ruleVisitor, [node], pluginContext);
+  BasePlugin.prototype.visitors = function() { };
+
+  // Generate a visitor handler function for a specific node type/event that
+  // will call the internal handlers and the rule handler for that node
+  // type/event in the correct order.
+  BasePlugin.prototype._wrapVisitorHandler = function(ruleHandler, internalHandlers) {
+    internalHandlers = internalHandlers || {};
+    var beforeHandlers = internalHandlers.before || [];
+    var afterHandlers = internalHandlers.after || [];
+    var plugin = this;
+    var i, ret;
+
+    return function() {
+      for (i = 0; i < beforeHandlers.length; i++) {
+        ret = beforeHandlers[i].apply(plugin, arguments);
+        if (ret === false) {
```

Comment:
This is **really** clever, and I love it!  Can we change the "special" value to be not `false` though?  Something like a module scoped `var REMOVE_HANDLER = {};` (and the handlers can just do `return REMOVE_HANDLER;`)...